src/GoTrueClient.ts

```diff
@@ -1847,6 +1847,14 @@ export default class GoTrueClient {
   ): {
     data: { subscription: Subscription }
   } {
+    if (callback.constructor.name === 'AsyncFunction') {
+      console.warn(
+        'Calling onAuthStateChange with an async function may cause a deadlock.'
+        + '  For a solution please see the docs: ' +
+        'https://supabase.com/docs/reference/javascript/auth-onauthstatechange'
+      )
+    }
```

Comment:
In and of itself there is no problem subscribing an async function, but the problem arises if you use another Supabase Client Library within, and that is almost impossible to detect reliably :(

If you want, you can try to find a way -- maybe setting a boolean on the object and checking it in any other auth-js method?

Reply:
Yep, I totally agree.  I couldn't think of any elegant way of dealing with this but wanted to help prevent someone else spending half a day getting confused by a race condition.  Perhaps this warning could only be displayed if the debug flag was set to true?